packages/node/warpsync/validation.go

```diff
@@ -0,0 +1,190 @@
+	m.startValidation()
+	defer m.endValidation()
+
+	ecChain = make(map[epoch.Index]epoch.EC)
+	ecRecordChain := make(map[epoch.Index]*epoch.ECRecord)
+	validPeers = set.NewAdvancedSet(m.p2pManager.AllNeighborsIDs()...)
+	neighborCommitments := make(map[epoch.Index]map[identity.ID]*neighborCommitment)
+
+	// We do not request the start nor the ending epoch, as we know the beginning (snapshot) and the end (tip received via gossip) of the chain.
+	startRange := start + 1
+	endRange := end - 1
+	for ei := endRange; ei >= startRange; ei-- {
+		m.requestEpochCommittment(ei)
+	}
+
+	epochToValidate := endRange
+	ecChain[start] = startEC
+	ecRecordChain[end] = epoch.NewECRecord(end)
+	ecRecordChain[end].SetPrevEC(endPrevEC)
+
+	for {
+		select {
+		case commitment, ok := <-m.commitmentsChan:
+			if !ok {
+				return nil, nil, nil
+			}
+			ecRecord := commitment.ecRecord
+			peerID := commitment.neighbor.Peer.ID()
+			commitmentEI := ecRecord.EI()
+			m.log.Debugw("read committment", "EI", commitmentEI, "EC", ecRecord.ComputeEC().Base58())
+			// Ignore invalid neighbor.
+			if !validPeers.Has(peerID) {
+				m.log.Debugw("ignoring invalid neighbor", "ID", peerID, "validPeers", validPeers)
+				continue
+			}
+
+			// Ignore committments outside of the range.
+			if commitmentEI < startRange || commitmentEI > endRange {
+				m.log.Debugw("ignoring committment outside of requested range", "EI", commitmentEI, "peer", peerID)
+				continue
+
+			}
+
+			// If we already validated this epoch, we check if the neighbor is on the target chain.
+			if commitmentEI > epochToValidate {
+				if ecRecordChain[commitmentEI].ComputeEC() != ecRecord.ComputeEC() {
+					m.log.Infof("ignoring commitment outside of the target chain", "peer", peerID)
+					validPeers.Delete(peerID)
+				}
+				continue
+			}
+
+			// commitmentEI <= epochToValidate
+			if neighborCommitments[commitmentEI] == nil {
+				neighborCommitments[commitmentEI] = make(map[identity.ID]*neighborCommitment)
+			}
+			neighborCommitments[commitmentEI][peerID] = commitment
+
+			// We received a committment out of order, we can evaluate it only later.
+			if commitmentEI < epochToValidate {
+				continue
+			}
+
+			// commitmentEI == epochToValidate
+			// Validate commitments collected so far.
+			for {
+				neighborCommitmentsForEpoch, received := neighborCommitments[epochToValidate]
+				// We haven't received commitments for this epoch yet.
+				if !received {
+					break
+				}
+
+				for peerID, epochCommitment := range neighborCommitmentsForEpoch {
+					if !validPeers.Has(peerID) {
+						continue
+					}
+					proposedECRecord := epochCommitment.ecRecord
+					if ecRecordChain[epochToValidate+1].PrevEC() != proposedECRecord.ComputeEC() {
+						m.log.Infof("ignoring commitment outside of the target chain", "peer", peerID)
+						validPeers.Delete(peerID)
```

Comment:
What happens if a neighbor didn't send us any ECRecords? Is it still seen as a valid peer for the epoch sync?